cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/DatastoreManager.java

```diff
@@ -57,12 +58,17 @@
  */
 public class DatastoreManager {
 
+    // Static map to manage DatastoreManager instances (one per directory path)
+    private static final ConcurrentMap<File, DatastoreManager> datastoreManagers = new
+            ConcurrentHashMap<File, DatastoreManager>();
+
     private final static String LOG_TAG = "DatastoreManager";
     private final static Logger logger = Logger.getLogger(DatastoreManager.class.getCanonicalName());
 
     private final String path;
 
-    private final Map<String, Datastore> openedDatastores = Collections.synchronizedMap(new HashMap<String, Datastore>());
+    /* This map should only be accessed inside synchronized(openDatastores) blocks */
```

Comment:
A `Collections.synchronizedMap` will synchronize all access to the map via the `Map` interface methods. This only guarantees that the elements are not modified during the operation so multiple callers see a consistent picture of the `Map`.
Caller 1: openDatastore("ds")
Caller 2: deleteDatastore("ds")
Caller 1 would get create the `Datastore` instance, and caller 2 would remove it from the Map and proceed to delete the files whilst caller 1 was also creating them. To avoid that problem we need to lock around the entire on disk action.
Once we have a `synchronized` block that extends around the on-disk work we can perform the `Map` access in the same block so there is no value in having a `Collections.synchronizedMap`. The locks are re-entrant so I don't think there is an issue with leaving it as a synchonizedMap if you prefer.
The real issue was the window presented by the double lock pattern, which could have released the lock for long enough (between the `contains` check and the `get`) for a `remove` operation to happen (as part of a delete) and result in the method returning `null`.